Issue 53: First pass at differenced AR

seabbs opened the pull request:
This PR is ready for review and closes #53

seabbs commented:
Any feedback before I push on?

SamuelBrand1 commented:
I'm still parsing this

SamuelBrand1 commented on EpiAware/src/abstract-functions.jl:
I think this can go in with abstract types? They aren't really abstract functions, they are the most general methods. I think of them as fallbacks.

SamuelBrand1 commented on EpiAware/src/latent-models.jl:
```diff
@@ -26,3 +21,90 @@ end
+    """
+    p::Int
+
+    function AR(damp_prior::Priors, var_prior::Distribution, init_prior::Priors)
```
I think making at least one constructors with kwargs might be nice?

seabbs replied:
Can you dummy out what you mean?

seabbs replied:
See #133

seabbs commented:
This is ready for review. 

I have added unit tests but given the blocker of #126 I think we need to wait for that for the documentation (which as noted is a serious issue). 

I have also opened a new PR for the differencing (#134 which I will address after this) and made an issue for the lack of consistency for positional args (#133).

SamuelBrand1 commented on EpiAware/src/latentmodels/autoregressive.jl:
In #130 I'm changing to `std_prior` which I think is better (agreed f2f) so probably worth changing here.

SamuelBrand1 commented on EpiAware/src/latentmodels/autoregressive.jl:
```diff
@@ -0,0 +1,53 @@
+struct AR <: AbstractLatentModel
```
Its usually better to use concrete types for fields of structs. This looks like you should make it a [parameteric](https://docs.julialang.org/en/v1/manual/types/#Parametric-Types) typed struct.

SamuelBrand1 reviewed (dismissed):
This looks great; love the self-referencing approach to getting generic AR/diff AR!

I think the main `AR` struct needs to be parametric, thats the only change I recommend.

seabbs commented:
I've updated w/ all suggestions. 

Note I have used `@with_kw` from parameters.jl vs the base implementation as it allows for assets to be placed inside the inner constructor and so makes the implementation cleaner. I have also removed the specification of `p` as this doesn't work well when using defaults as it starts to conflict with the inner constructor.

seabbs commented:
@SamuelBrand1 pinging for review

SamuelBrand1 commented:
> @SamuelBrand1 pinging for review

Reviewing this now

SamuelBrand1 commented on EpiAware/src/latentmodels/autoregressive.jl:
```diff
@@ -0,0 +1,46 @@
+@doc raw"
+The autoregressive (AR) model struct.
+"
+@with_kw struct AR{D <: Priors, S <: Distribution, I <: Priors} <: AbstractLatentModel
```
I like this use of `Parameters.@with_kw`. Happy to have `Parameters.jl` as a dep.

SamuelBrand1 commented on EpiAware/test/test_autoregressive.jl:
```diff
@@ -0,0 +1,62 @@
+    @test ar_process.init_prior == init_prior
+end
+
+@testitem "Test AR defaults" begin
```
Could you test this for creating a vector length `damp` variable?

seabbs replied:
Good spot your right it isn't. I will follow the pattern in #134 and reping for a review.

SamuelBrand1 reviewed (dismissed):
LGTM, only concern is that the method isn't robust to generating vector shaped `damp` or `ar_init`.

If this gets tested and works then I'm happy

seabbs commented:
This should be updated now following the approach suggested in #134 and rebased on main. Note I had to copy `docstring`z templates everywhere due to submodules (a good downside we should have flagged) but I think the duplication isn't that bad as its something that gets touched rarely. 

Moving on to #134 now.

SamuelBrand1 commented on EpiAware/src/EpiLatentModels/utils.jl:
```diff
@@ -0,0 +1,6 @@
+function _expand_dist(dist::Vector{D} where {D <: Distribution})
```
Nice addition

SamuelBrand1 reviewed (approved):
I like this.
